architecture/ambient/lifecycle.md

```diff
@@ -0,0 +1,136 @@
+# Ztunnel Lifecyle
+
+This document describes the lifecycle of various aspects of Ztunnel, and its relationship to workloads and the CNI.
+
+At a high level, our goal is to provide complete connectivity to a workload throughout its entire lifetime.
+Failing to do so can be an availability risk (if we deny traffic that should succeed) or a security risk (if we allow traffic that should be denied).
+
+## High level overview
+
+At a high level, the relevant components look as such:
+
+```mermaid
+flowchart TD
+    CNIP["CNI Plugin"]
+    CNI["CNI Agent"]
+    Z["Ztunnel"]
+    Pods["Pods"]
+    CNI --Programs--> Z
+    CNI --Watches--> Pods
+    Z --Runs Within--> Pods
+    CNIP --Calls--> CNI
+```
+
+The CNI Plugin is a binary installed as a [CNI plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/) on the node.
+Kubernetes is responsible for invoking this when a Pod is being started (before the containers run).
+When this occurs, the plugin will call out to the CNI Agent to program the network.
+This includes setting up networking rules both within the pod network namespace, and also on the host network.
+
+An alternative flow is when a pod is enabled after startup.
+In this case, the CNI Agent is watching for Pod events from the API server directly and performing the same setup.
+
```

Comment:
This deserves a lot more details - for developers and admins - it is not even close to the same setup. 

Watching the pods doesn't provide the network namespace - we perform a duct-taped search in /proc with some regex to guess which processes are pods and what are their namespaces, hoping the CRI layer is one that happens to generate ( and guarantee) a particular naming pattern that was reverse engineered from undocumented and unsupported behaviors of common CRIs.

I think it is critical for any dev ( or operator ) to understand this part - it's one of the most tricky areas.